src/util/codec/convert.rs

```diff
@@ -133,13 +132,97 @@ fn get_valid_float_prefix(s: &str) -> &str {
         }
     }
 
-    if valid_len == 0 { "0" } else { &s[..valid_len] }
+    try!(handle_truncate(ctx, valid_len < s.len()));
+    if valid_len == 0 {
+        Ok("0")
+    } else {
+        Ok(&s[..valid_len])
+    }
 }
 
+/// It converts a valid float string into valid integer string which can be
+/// parsed by `i64::from_str`, we can't parse float first then convert it to string
+/// because precision will be lost.
+fn float_str_to_int_string<'a, 'b: 'a>(valid_float: &'b str) -> Result<Cow<'a, str>> {
+    let mut dot_idx = None;
+    let mut e_idx = None;
+    let mut int_cnt: i64 = 0;
+    let mut digits_cnt: i64 = 0;
+
+    for (i, c) in valid_float.chars().enumerate() {
+        match c {
+            '.' => dot_idx = Some(i),
+            'e' | 'E' => e_idx = Some(i),
+            '0'...'9' => {
+                if e_idx.is_none() {
+                    if dot_idx.is_none() {
+                        int_cnt += 1;
+                    }
+                    digits_cnt += 1;
+                }
+            }
+            _ => (),
+        }
+    }
+
+    if dot_idx.is_none() && e_idx.is_none() {
+        return Ok(Cow::Borrowed(valid_float));
+    }
+
+    if e_idx.is_none() {
+        if int_cnt == 0 {
+            return Ok(Cow::Borrowed("0"));
+        }
+        return Ok(Cow::Borrowed(&valid_float[..dot_idx.unwrap()]));
+    }
+
+    let exp = box_try!((&valid_float[e_idx.unwrap() + 1..]).parse::<i64>());
+    if exp > 0 && int_cnt > (i64::MAX - exp) {
+        // (exp + inc_cnt) overflows MaxInt64.
+        // TODO: refactor errors
+        return Err(box_err!("[1264] Data Out of Range"));
+    }
+    if int_cnt + exp <= 0 {
+        return Ok(Cow::Borrowed("0"));
+    }
+
+    let mut valid_int = String::from(&valid_float[..e_idx.unwrap()]);
+    if dot_idx.is_some() {
+        valid_int.remove(dot_idx.unwrap());
+    }
+
+    let extra_zero_count = exp + int_cnt - digits_cnt;
+    if extra_zero_count > MAX_ZERO_COUNT {
+        // Return overflow to avoid allocating too much memory.
+        // TODO: refactor errors
+        return Err(box_err!("[1264] Data Out of Range"));
+    }
+
+    if extra_zero_count >= 0 {
+        valid_int.extend((0..extra_zero_count).map(|_| '0'));
+    } else {
+        let len = valid_int.len();
+        if extra_zero_count >= len as i64 {
+            return Ok(Cow::Borrowed("0"));
+        }
+        valid_int.truncate((len as i64 + extra_zero_count) as usize);
+    }
+
+    while valid_int.starts_with('0') && valid_int.len() > 1 {
```

Comment:
This is handled properly in when converting string to int. And this while loop is not efficient enough, it will copy all the data in every loop.